File: graph-service-examples/nodejs/load-balancer/use-balancer.js
const gremlin = require('gremlin');
const { traversal } = gremlin.process;
const { RoundRobinClientRemoteConnection } = require('./load-balancer');

const endpoints = ['localhost:8181', 'localhost:8182', 'localhost:8183'];

const lbLogger = (level, msg) => {
    const ts = new Date().toISOString();
    console.log(`${ts} RoundRobinClientRemoteConnection ${level.toUpperCase()}: ${msg}`);
};

(async () => {
    const rr = new RoundRobinClientRemoteConnection(
        endpoints,
        'g',
        2,
        lbLogger // debug logger
    );

    try {
        const g = traversal().withRemote(rr);

        const user1 = await g.addV('User').property('userId', 'U1').property('name', 'Alice').property('age', 30).next();
        const user2 = await g.addV('User').property('userId', 'U2').property('name', 'Bob').property('age', 25).next();
        const user3 = await g.addV('User').property('userId', 'U3').property('name', 'Charlie').property('age', 35).next();

        const account1 = await g.addV('Account').property('accountId', 'A1').property('balance', 1000).next();
        const account2 = await g.addV('Account').property('accountId', 'A2').property('balance', 500).next();
        const account3 = await g.addV('Account').property('accountId', 'A3').property('balance', 750).next();

        await g.addE('owns').from_(user1.value).to(account1.value).iterate();
        await g.addE('owns').from_(user2.value).to(account2.value).iterate();
        await g.addE('owns').from_(user3.value).to(account3.value).iterate();

        let i = 1;
        while (true) {
            try {
                const list = await g.V().limit(i).toList();
                console.log(`Fetched ${list.length} vertices`);
            } catch (e) {
                console.warn('Traversal failed:', e.message || e);
            }
            await new Promise((r) => setTimeout(r, 3000));
            i += 1;
        }
    } catch (e) {
        console.error('Fatal error:', e);
    } finally {
        process.on('SIGINT', async () => {
            await rr.close();
            console.log('Load balancer closed. Goodbye!');
            process.exit(0);
        });
    }
})();

File: graph-service-examples/python/load_balancer/use_balancer.py
# Copyright 2022-2026 Aerospike, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import sys
from time import sleep
from gremlin_python.process.anonymous_traversal import traversal
from load_balancer import RoundRobinClientRemoteConnection

# Logging Configurations
root = logging.getLogger()
root.setLevel(logging.INFO)
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(
    logging.Formatter("%(asctime)s %(name)s %(levelname)s: %(message)s")
)
root.addHandler(handler)
lb_logger = logging.getLogger("RoundRobinClientRemoteConnection")
lb_logger.setLevel(logging.DEBUG) # Change this if you want less verbose output
logging.getLogger("gremlin_python").setLevel(logging.WARNING)
logging.getLogger("asyncio").setLevel(logging.WARNING)

endpoints = [
    "localhost:8181",
    "localhost:8182",
    "localhost:8183"
]

rr_conn = RoundRobinClientRemoteConnection(endpoints, traversal_source="g",
                                           log_level=logging.DEBUG, logger=lb_logger )
g = traversal().with_remote(rr_conn)
results = []

user1 = g.add_v("User").property("userId", "U1").property("name", "Alice").property("age", 30).next()
user2 = g.add_v("User").property("userId", "U2").property("name", "Bob").property("age", 25).next()
user3 = g.add_v("User").property("userId", "U3").property("name", "Charlie").property("age", 35).next()

account1 = g.add_v("Account").property("accountId", "A1").property("balance", 1000).next()
account2 = g.add_v("Account").property("accountId", "A2").property("balance", 500).next()
account3 = g.add_v("Account").property("accountId", "A3").property("balance", 750).next()

g.add_e("owns").from_(user1).to(account1).next()
g.add_e("owns").from_(user2).to(account2).next()
g.add_e("owns").from_(user3).to(account3).next()

try:
    i = 1
    while True:
        try:
            results = g.V().limit(i).to_list()
            sleep(3)
            i += 1
        except Exception as e:
            logging.warning("Traversal failed: %s", e)
except KeyboardInterrupt:
    print("\nReceived interrupt, shutting down load balancer loop.")
finally:
    rr_conn.close()
    print("Load balancer closed. Goodbye!")